### DISCUSSION TITLE
custom source: get_keyword_pattern vs get_trigger_characters

### DISCUSSION THREAD

**Discussion by vicecaycedo** [hrsh7th/nvim-cmp#971] [Q&A] [answered]
I'm working on creating a custom source and I'm confused as to what get_keyword_pattern and get_trigger_characters do.

For context, I'm working on a cmp source that only suggests items when "b/" is typed. Setting the keyword_pattern to ‘b’ and the trigger character to ‘/’ seems to work, but the source:complete function gets called every time ‘b’ is typed. Ideally, this function would only be called when ‘b/’ is typed.

I appreciate the help! Thanks!

**Comment by vicecaycedo** [comment, marked as answer]
The solution is to have the keyword pattern be 'b/' and omitting the trigger characters. The other important part is that the response item labels have to begin with the keyword pattern, e.g. 'b/' in this case.

**Reply by hrsh7th** [reply]
You are right.

The trigger character is needed to support the thing like `object.|` situation.

Typically, the completion engine does not invoke completion with less than 1 keyword characters.
But in the situation of `object.|`, we might hope to suggest all of the object's methods.

In this case, you should add `.` as trigger characters.

**Reply by nyngwang** [reply]
Oh @vicecaycedo really thank you for asking this, I'm building a source too!

**Reply by nyngwang** [reply]
@hrsh7th:

- What if I want to trigger it on every keystroke, so those items will always appear in the menu?
- What if I don't want to trigger it until I press some keymap like `<C-g>`?
- I just saw some source has both `get_keyword_pattern`, `get_trigger_characters` missing, and it use `callback({ isIncomplete = true })` instead. What is that?
